### PR TITLE
DB-11966: set selectivityEstimationIncludingSkewedDefault for all servers

### PR DESCRIPTION
**Pull request by jyuanca**
## Short Description
`selectivityEstimationIncludingSkewedDefault` was defined as a service property, which cannot be set globally by `syscs_set_global_database_property`.

## Long Description
`selectivityEstimationIncludingSkewedDefault` was defined as a service property, which cannot be set globally by `syscs_set_global_database_property`. To fix it, remove it from service properties, and fix the code to retrieve it from service property.

## How to test

Run 

`call syscs_util.syscs_set_global_database_property('derby.database.selectivityEstimationIncludingSkewedDefault', 'true');`

to set it to true. Verify it is set for all servers

```
call syscs_util.syscs_get_global_database_property('derby.database.selectivityEstimationIncludingSkewedDefault');
HOST_NAME                     |PROPERTY_VALUE
-----------------------------------------------
Juns-MacBook-Pro-2.local:1527 | true  
Juns-MacBook-Pro-2.local:1528 | true

2 rows selected
```

### CODE REVIEW

**Review by arnaud-splice — APPROVED** [review]

**Review by carolp-503 — APPROVED** [review]
+1

**Comment by jyuanca** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/DataDictionaryImpl.java [hunk 1]
```diff
@@ -533,7 +533,6 @@ public void boot(boolean create,Properties startParams) throws StandardException
                     }
                 }
             }
-            startParams.setProperty("catalogVersion", dictionaryVersion.toString());
```
Comment (dgomezferro) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/DataDictionaryImpl.java:536]: Why is this removed?

Comment (jyuanca) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/DataDictionaryImpl.java:536]: dictionaryVersion is always derby version 10.x. We should set splice build version. It is set somewhere else

File: splice_machine/src/main/java/com/splicemachine/derby/impl/SpliceService.java [hunk 2]
```diff
@@ -61,8 +62,10 @@ public Properties getServiceProperties(String serviceName, Properties defaultPro
 			Set<String> properties = propertyManager.listProperties();
 //			List<String> children = ZkUtils.getChildren(zkSpliceDerbyPropertyPath, false);
 			for (String property: properties) {
-				String value = propertyManager.getProperty(property);
-				service.setProperty(property, value);
+				if (!property.equals(Property.SELECTIVITY_ESTIMATION_INCLUDING_SKEWED)) {
```
Comment (martinrupp) [inline splice_machine/src/main/java/com/splicemachine/derby/impl/SpliceService.java:65]: please add a comment why we need special handling of SELECTIVITY_ESTIMATION_INCLUDING_SKEWED here. if this is because SELECTIVITY_ESTIMATION_INCLUDING_SKEWED is now not stored in propertyManager anymore, then i think we should have sth like `PropertyUtil.deletedPropertiesList`

File: splice_machine/src/main/java/com/splicemachine/derby/impl/store/access/PropertyConglomerate.java [hunk 3]
```diff
@@ -119,6 +118,10 @@ public PropertyConglomerate(TransactionController tc, boolean create, Properties
         try {
             Set<String> props = propertyManager.listProperties();
             for (String child: props) {
+                if (child.equals(Property.SELECTIVITY_ESTIMATION_INCLUDING_SKEWED)) {
```
Comment (martinrupp) [inline splice_machine/src/main/java/com/splicemachine/derby/impl/store/access/PropertyConglomerate.java:121]: `PropertyUtil.deletedPropertiesList.contains(child)`

**Review by dgomezferro — APPROVED** [review]

**Review by martinrupp — CHANGES_REQUESTED** [review]

File: hbase_sql/src/main/java/com/splicemachine/derby/impl/sql/ZkPropertyManager.java [hunk 4]
```diff
@@ -67,8 +67,14 @@ public String getProperty(String propertyName) throws StandardException{
...
     public void addProperty(String propertyName,String propertyValue) throws StandardException{
+        String path = zkSpliceDerbyPropertyPath + "/" + propertyName;
         try{
-            ZkUtils.safeCreate(zkSpliceDerbyPropertyPath+"/"+propertyName,Bytes.toBytes(propertyValue),ZooDefs.Ids.OPEN_ACL_UNSAFE,CreateMode.PERSISTENT);
+            if (ZkUtils.getRecoverableZooKeeper().exists(path, false) == null) {
+                ZkUtils.safeCreate(path, Bytes.toBytes(propertyValue), ZooDefs.Ids.OPEN_ACL_UNSAFE, CreateMode.PERSISTENT);
+            }
+            else {
+                ZkUtils.setData(path, Bytes.toBytes(propertyValue), -1);
+            }
```
Comment (hatyo) [inline hbase_sql/src/main/java/com/splicemachine/derby/impl/sql/ZkPropertyManager.java:77]: This is not an atomic operation anymore, for a good reason, as `safeCreate` does [not](https://github.com/sfines/menagerie/blob/master/src/main/java/org/menagerie/ZkUtils.java#L206) alter the data if the path exist.

Does it make sense to make it atomic, e.g. by obtaining a distributed lock on the `path` itself?

By the way, is setting a property a transactional operation? at first glance it seems not, right, from the looks of this method, it looks like last-writer-wins, causing this anomaly:

```
client 1                                                       client 2

autocommit off                                                 autocommit off
call syscs_util.syscs_set_global_database_property('a','b');
call syscs_util.syscs_set_global_database_property('c','e');
                                                               call syscs_util.syscs_set_global_database_property('a','z');
commit;
                                                               commit;
```

End result:
```
PropertyName,PropertyValue
a,z
c,e
```

**Review by hatyo — COMMENTED** [review]
PR looks good me given that @martinrupp's comment is addressed.

My comment was meant mainly to spark a discussion about transactionality of properties, and maybe revisiting the property model.

Let us have a follow up discussion to revisit the following topics:
- Is it expected that global database properties are not transactional? if so, is this properly documented?
- Should we encourage devs to think about having the property defined in a session if possible, and only add it as a database property if and only if the property rarely changes and is meant to change the behavior of the entire system.
- global database properties do affect all database in a multi-db setup, should we have a database-specific layer allowing us to e.g. set min. retention period for DB1 to be 4 days and for DB2 to be 15 days?
- some properties look like a session property but could have cross-session side-effects, e.g. it is possible to set a property that affect a statement's plan, if that plan is cached and used by other sessions then this property has a cross-session side-effect. Should we revisit this model? (e.g. cache statements affected by such properties in a special session-specific statement-cache?)

File: hbase_sql/src/main/java/com/splicemachine/derby/impl/sql/ZkPropertyManager.java (hunk 4 above)
Comment (jyuanca) [reply hbase_sql/src/main/java/com/splicemachine/derby/impl/sql/ZkPropertyManager.java:77]: property is not transactional

**Comment by jyuanca** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by jyuanca** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Comment by jyuanca** [comment]
@martinrupp please review latest changes.

**Review by martinrupp — APPROVED** [review]

File: hbase_sql/src/main/java/com/splicemachine/derby/impl/sql/ZkPropertyManager.java (hunk 4 above)
Comment (jyuanca) [reply hbase_sql/src/main/java/com/splicemachine/derby/impl/sql/ZkPropertyManager.java:78]: If the path exists, the data is modified.

**Comment by jyuanca** [comment]
Filed  DB-12091 to followup Youssef's comments

File: hbase_sql/src/main/java/com/splicemachine/derby/impl/sql/ZkPropertyManager.java (hunk 4 above)
Comment (hatyo) [reply hbase_sql/src/main/java/com/splicemachine/derby/impl/sql/ZkPropertyManager.java:78]: > Filed DB-12091 to followup Youssef's comments

Thank you @jyuanca

**Review by hatyo — APPROVED** [review]
